src/expr/src/relation/mod.rs

```diff
@@ -455,6 +455,162 @@ impl RelationExpr {
         self.typ().column_types.len()
     }
 
+    /// Recursively descend two expressions until differences are found.
+    ///
+    /// The resulting sequence contains the first moments at which differences were observed.
+    /// There may be multiple elements in the list as the differences may only be observed
+    /// when descending independent tree paths.
+    pub fn tree_differences(&self, other: &RelationExpr) -> Vec<(RelationExpr, RelationExpr)> {
+        match (self, other) {
+            (RelationExpr::Constant { .. }, RelationExpr::Constant { .. }) if self == other => {
+                Vec::new()
+            }
+            (RelationExpr::Get { .. }, RelationExpr::Get { .. }) if self == other => Vec::new(),
+            (
+                RelationExpr::Let {
+                    id: i1,
+                    value: v1,
+                    body: b1,
+                },
+                RelationExpr::Let {
+                    id: i2,
+                    value: v2,
+                    body: b2,
+                },
+            ) if i1 == i2 => {
+                let mut result = Vec::new();
```

Comment:
This method is meant to be used to present differences in plans, rather than compare things within one plan. So, e.g. if you had a transform that changed the value of a let (maybe pushed a predicate down into it) this would notice that and inform you.

The method should probably be removed though (I thought I had, but I think I lost it in a rebase that I did).